README.md
README
======

This project aims at providing a network namespace api for GNU/Linux server.

# DOC
The swagger documentation is available at `http://localhost/api/vX/doc/  `
where X is the version number of the api  
ex: http://localhost/api/v1/doc/

CHANGELOG.md
CHANGELOG
=========

# v0.1

## NEW
- Initiation of the project
- You can browse the list of the network namespace existing on the server
- You can create a network namespace

tests/tests.py
from __future__ import absolute_import
import unittest
import requests
from flask import Flask
from apps.apiv1 import blueprint as apiv1


class TestWSMethods(unittest.TestCase):
    def setUp(self):
        self.app = Flask(__name__)
        self.app.register_blueprint(apiv1)

    def test_get_netns(self):
        self.res = requests.get('http://127.0.0.1:5000/api/v1/netnamespaces/')
        self.assertEqual(self.res.status_code, 200)


if __name__ == '__main__':
    unittest.main()


namespaces/netns.py
from flask_restplus import Namespace, Resource, abort
from pyroute2 import NetNS, netns
from models.model_netns import NetnsDao, model_netns
from models.model_interface import model_interface

# Create a namespace
nsnetns = Namespace('netnamespaces', description='Network namespace')

# Register models
nsnetns.models[model_netns.name] = model_netns
nsnetns.models[model_interface.name] = model_interface


# Create routes to work with the resource
@nsnetns.route('/')
class NetNamespaces(Resource):
    @nsnetns.doc('List network namespaces')
    @nsnetns.response(200, 'Success')
    def get(self):
        return netns.listnetns()


@nsnetns.route('/<string:nspath>')
@nsnetns.doc(params={'nspath': 'a network namespace name'})
class NetNamespace(Resource):

    @nsnetns.marshal_with(model_netns)
    @nsnetns.doc('Retrieve a network namespace detail')
    @nsnetns.response(200, 'Network namespace detail')
    @nsnetns.response(404, 'Namespace does not exist')
    def get(self, nspath):
        if nspath in netns.listnetns():
            return NetnsDao(nspath)
        else:
            abort(404, "Namespace " + nspath + " does not exist")

    def put(self, nspath):
        if nspath not in netns.listnetns():
            abort(404, "Namespace not found")
        else:
            pass

    @nsnetns.doc('Create a network namespace')
    @nsnetns.response(200, 'Network namespace successfully created')
    @nsnetns.response(500, 'Network namespace already exists')
    def post(self, nspath):
        if nspath in netns.listnetns():
            abort(500, "Namespace already exists")
        else:
            mynet = NetNS(nspath)
            mynet.close()
            return("Namespace successfully created")

    @nsnetns.doc('Delete a network namespace')
    @nsnetns.response(200, 'Network namespace successfully removed')
    @nsnetns.response(404, 'Network namespace not found')
    def delete(self, nspath):
        if nspath not in netns.listnetns():
            abort(404, "Namespace not found")
        else:
            mynet = NetNS(nspath)
            mynet.remove()
            mynet.close()
            return("Namespace successfully removed")


@nsnetns.route('/<string:nspath>/interfaces')
class NetNsIp(Resource):
    @nsnetns.marshal_with(model_interface)
    @nsnetns.doc('List namespace interfaces')
    @nsnetns.response(200, 'Returns the list of interface with their ip')
    @nsnetns.response(404, 'Namespace does not exist')
    def get(self, nspath):
        if nspath in netns.listnetns():
            return NetnsDao(nspath).get_interfaces_dao()
        else:
            abort(404, "Namespace does not exist")

    def put(self):
        pass

    def post(self):
        pass

    def delete(self):
        pass


models/model_netns.py
from flask_restplus import Model, fields
from pyroute2 import IPDB, NetNS
from model_interface import InterfaceDao

# Create a model for my resource
model_netns = Model('Netns_model', {
    'name': fields.String,
    'interfaces_name': fields.List(fields.String)
})


class NetnsDao():
    '''
    The NetnsDao class provides the short representation of a network namespace

    :attr name: the name of the namespace
    :attr interfaces: the list of interfaces in the namespace
    :param nspath: the name of the namespace
    :type nspath: string
    '''
    def __init__(self, nspath):
        self.name = nspath
        self.interfaces_name = []
        for i in self.get_interfaces_dao():
            self.interfaces_name.append(i.ifname)

    def get_interfaces_dao(self):
        '''
        This method retrieves the list of interfaces for the network namespace.
        It will returns a list of InterfaceDao object.

        :returns: list of interfaces in the namespace
        :rtype: list of :class:InterfaceDao
        '''
        interfaces = []
        with IPDB(nl=NetNS(self.name)) as ipdb:
            for i in ipdb.interfaces.iteritems():
                interface = InterfaceDao(i[1])
                interfaces.append(interface)
        return interfaces


models/model_interface.py
from flask_restplus import Model, fields

model_interface = Model('Interface_model', {
    'ifname': fields.String,
    'ip': fields.String,
    'status': fields.String
})


class InterfaceDao():
    '''
    The InterfaceDao class provides the short representation of
    a network interfaces

    :attr ifname: the name of the interface
    :attr ip: the interface IPV4
    :attr status: the operational state of the interface
    :param ipdb_interface: a network interface
    :type ipdb_interface: an IPDB interface object
                                <class 'pyroute2.ipdb.interface.Interface'>
    '''
    def __init__(self, ipdb_interface):
        try:
            self.ifname = ipdb_interface.ifname
        except:
            pass
        try:
            self.ip = ipdb_interface.ipaddr[0]['address']
        except:
            pass
        try:
            self.status = ipdb_interface.operstate
        except:
            pass


namespaces/netnsdao.py
from flask_restplus import Namespace, Resource, fields, abort
from pyroute2 import IPDB, NetNS, netns

# Create a namespace
nsnetns = Namespace('netnamespaces', description='Network namespace')

# Create a model for my resource
model_netns = nsnetns.model('Netns_model', {
    'netns': fields.String,
})

model_ip = nsnetns.model('Interface_model', {

})


# Create routes to work with the resource
@nsnetns.route('/')
class NetNamespaces(Resource):
    @nsnetns.doc('List network namespaces')
    @nsnetns.response(200, 'Success')
    def get(self):
        namespaces = netns.listnetns()
        return namespaces


@nsnetns.route('/<string:nspath>')
@nsnetns.doc(params={'nspath': 'a network namespace name'})
class NetNamespace(Resource):
    @nsnetns.marshal_with(model_netns)
    @nsnetns.doc('Retrieve a network namespace detail')
    @nsnetns.response(200, 'Success')
    @nsnetns.response(404, 'Namespace does not exist')
    def get(self, nspath):
        if nspath in netns.listnetns():
            namespace = NetNS(nspath)
            return namespace
        else:
            abort(404, "Namespace " + nspath + " does not exist")

    def put(self, nspath):
        if nspath not in netns.listnetns():
            abort(404, "Namespace not found")
        else:
            pass

    @nsnetns.doc('Create a network namespace')
    @nsnetns.response(200, 'Network namespace successfully created')
    @nsnetns.response(500, 'Network namespace already exists')
    def post(self, nspath):
        if nspath in netns.listnetns():
            abort(500, "Namespace already exists")
        else:
            mynet = NetNS(nspath)
            mynet.close()
            return("Namespace successfully created")

    @nsnetns.doc('Delete a network namespace')
    @nsnetns.response(200, 'Network namespace successfully removed')
    @nsnetns.response(404, 'Network namespace not found')
    def delete(self, nspath):
        if nspath not in netns.listnetns():
            abort(404, "Namespace not found")
        else:
            mynet = NetNS(nspath)
            mynet.remove()
            mynet.close()
            return("Namespace successfully removed")


@nsnetns.route('/<string:nspath>/ips')
@nsnetns.doc(params={'nspath': 'a network namespace name'})
class NetNsIp(Resource):
    @nsnetns.doc('List namespace ip')
    @nsnetns.response(404, 'Namespace does not exist')
    def get(self, nspath):
        if nspath in netns.listnetns():
            interface = ()
            ips = []
            ipdb = IPDB(nl=NetNS(nspath))
            for i in ipdb.interfaces.iteritems():
                interface = (i[0], i[1].ipaddr[0]['address'])
                ips.append(interface)
            return ips
        else:
            abort(404, "Namespace " + nspath + " does not exist")

    def put(self):
        pass

    def post(self):
        pass

    def delete(self):
        pass
